microprofile/WFLY-13640_MicroProfile_Reactive_Messaging.adoc

```diff
@@ -0,0 +1,183 @@
+
+There are several libraries all implementing the http://www.reactive-streams.org[Reactive Streams] specification. One new implementation is https://github.com/smallrye/smallrye-mutiny[SmallRye Mutiny] which is used under the hood by the SmallRye implementations of RSO and Reactive Messaging. We already use https://github.com/ReactiveX/RxJava/tree/2.x[RxJava 2] as an internal (i.e. it is a private module) dependency for the runtime functionality of some existing WildFly extensions. However, these should not be exposed to users at this time. We should settle on the implementation of RSO, which, although has less functionality than the other two mentioned, is a MicroProfile specification.
+
+In addition for more 'finite' sets of data, Java's CompletionStage is supported.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFLY-13640
+* https://issues.redhat.com/browse/EAP7-1405
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-1518[Quickstart]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:jstourac@redhat.com[Jan Stourac]
+* mailto:okotek@redhat.com[Ondrej Kotek]
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE.
+// Discuss with QE during the Kickoff state to decide this
+* [ ] Engineering
+
+* [X] QE
+
+=== Affected Projects or Components
+The changes will be made in WildFly adding a subsystem for each spec.
+
+The functionality of these subsystems will be provided by the SmallRye implementations and their dependencies.
+
+In addition for OpenShift we need to make sure that provides out-of-the box integration with the Kafka managed service (by leveraging integration at the operator level). This  integration is scheduled post-WildFly 23. My current understanding is that we should be able to use environment variables and system properties to indicate to the user application the address and ports of the Kafka managed service.
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+It must be possible to:
+
+* Deploy a Reactive Messaging application into WildFly
+* Interact with Kafka streams via the Reactive Messaging 1.0 APIS
+
+** MicroProfile Config will be used to configure the connections to the Kafka streams.
+
+*** It should be possible to override (or supplement?) parts of the MicroProfile Config via system properties or environment variables for integration with the Kafka managed service.
```

Comment:
@jmesnil As discussed yesterday.
I think probably 'supplement' is better since things like the commit strategy etc. should be set via the user application's config. IP or port of Kafka sounds more like external stuff.

If we want to override say the ports, then I believe we will need to do something, as IIRC the deployment microprofile-config.properties normally override the stuff coming from env vars and system properties. In that case we would need to do something to turn that on its head somewhat (possibly something like a ConfigSource looking for system properties and env vars following a certain format, returning them as properties and using a higher ordinal than the deployment microprofile-config.properties one).

Reply:
That's not correct: env vars have higher ordinal than microprofile-config.properties (https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/configsources.asciidoc#default-configsources)

So, the deployment can provide "sensible" default value in its microprofile-config.properties  but if we set an env var for this config property, it will take precedence

Reply:
@kabir @jmesnil So, given the above should the wording of this bullet become:

* Because MicroProfile Config supports systems properties or environmental variable as default ConfigSources with higher ordinals than is given to a microprofile-config.properties packaged in the application, it is possible for system that control the server launch to use system properties or environment variables to configure integration with the Kafka managed service.

Reply:
Incorporated Brian's suggestion. @jstourac FYI